File: backend/app/models/session.py
from __future__ import annotations

from typing import Any

from pydantic import Field, field_validator

from app.models.base import RequestModel, ResponseModel
from app.models.enums import (
    AcquisitionMode,
    AnalysisScope,
    AgentBootstrapState,
    DeviceType,
    RecoveryState,
    Scenario,
    SessionStatus,
)


class ParticipantIdentity(ResponseModel):
    """Identitas peserta seleksi yang terikat ke sesi akuisisi."""

    full_name: str = ""
    registration_no: str = ""
    nik: str | None = None
    organization: str | None = None


class ParticipantInput(RequestModel):
    full_name: str = Field(min_length=1, max_length=128)
    registration_no: str = Field(min_length=1, max_length=64)
    nik: str | None = Field(default=None, max_length=32)
    organization: str | None = Field(default=None, max_length=128)

    @field_validator("full_name", "registration_no", "nik", "organization", mode="before")
    @classmethod
    def strip_text(cls, value: Any) -> Any:
        if isinstance(value, str):
            cleaned = value.strip()
            return cleaned or None
        return value

    @field_validator("full_name", "registration_no")
    @classmethod
    def require_non_empty(cls, value: Any) -> Any:
        if not value:
            raise ValueError("wajib diisi")
        return value

    @field_validator("nik")
    @classmethod
    def nik_must_be_16_digits(cls, value: Any) -> Any:
        if value is None:
            return None
        text = str(value).strip()
        if not text:
            return None
        if not text.isdigit() or len(text) != 16:
            raise ValueError("NIK harus 16 digit angka")
        return text


class StartSessionRequest(RequestModel):
    device_id: str | None = Field(default=None, max_length=512)
    device_type: DeviceType = DeviceType.ANDROID
    mode: AcquisitionMode = AcquisitionMode.QUICK
    analysis_scope: AnalysisScope = AnalysisScope.DEVICE
    device_sources: list[str] = Field(default_factory=list)
    social_targets: list[str] = Field(default_factory=list)
    scenario: Scenario = Scenario.LULUS
    file_count: int = Field(default=1200, ge=1, le=1_000_000)
    label: str | None = Field(default=None, max_length=256)
    participant: ParticipantInput
    force_simulated: bool = False
    review_candidates: bool = False

    @field_validator("device_id", "label", mode="before")
    @classmethod
    def empty_string_as_none(cls, value: Any) -> Any:
        if isinstance(value, str) and not value.strip():
            return None
        return value

    def analysis_plan(self):
        from app.acquisition.analysis_plan import build_analysis_plan

        return build_analysis_plan(
            scope=self.analysis_scope,
            device_sources=self.device_sources,
            social_targets=self.social_targets,
        )


class UpdateParticipantRequest(RequestModel):
    participant: ParticipantInput


class AuthorizeRequest(RequestModel):
    note: str | None = Field(default=None, max_length=4000)


class SessionProgress(ResponseModel):
    phase: SessionStatus = SessionStatus.PENDING
    percent: float = Field(default=0, ge=0, le=100)
    message: str = "Menunggu"
    files_listed: int = Field(default=0, ge=0)
    files_pulled: int = Field(default=0, ge=0)
    files_indexed: int = Field(default=0, ge=0)
    files_analyzed: int = Field(default=0, ge=0)
    findings_count: int = Field(default=0, ge=0)
    throughput_files_per_sec: float = Field(default=0, ge=0)
    live_analysis_ms: float = Field(default=0, ge=0)
    acquisition_method: str | None = None
    hits_l1: int = Field(default=0, ge=0)
    hits_l2: int = Field(default=0, ge=0)
    hits_l3: int = Field(default=0, ge=0)
    hits_l4: int = Field(default=0, ge=0)
    hits_ocr: int = Field(default=0, ge=0)
    hits_asr: int = Field(default=0, ge=0)
    authorized_by: str | None = None
    authorized_at: str | None = None
    authorize_note: str | None = None
    bootstrap_state: AgentBootstrapState | None = None
    agent_state: str | None = None
    agent_version: str | None = None
    agent_api_version: str | None = None
    agent_install_action: str | None = None
    agent_retryable: bool | None = None
    agent_error_category: str | None = None
    runtime_permissions: dict[str, str] | None = None
    special_access: dict[str, str] | None = None
    selection_state: str | None = None
    selection_revision: int | None = Field(default=None, ge=1)
    selection_fingerprint: str | None = Field(default=None, pattern=r"^[0-9a-f]{64}$")
    selection_policy_version: str | None = None
    selection_policy_fingerprint: str | None = Field(
        default=None,
        pattern=r"^[0-9a-f]{64}$",
    )
    selection_candidates: int | None = Field(default=None, ge=0)
    selection_selected: int | None = Field(default=None, ge=0)
    recovery_state: RecoveryState | None = None
    recovery_mode: AcquisitionMode | None = None
    recovery_candidates: int | None = Field(default=None, ge=0)
    recovery_captured: int | None = Field(default=None, ge=0)
    recovery_bytes: int | None = Field(default=None, ge=0)
    recovery_warning_count: int | None = Field(default=None, ge=0)
    recovery_duration_ms: float | None = Field(default=None, ge=0)
    recovery_cache_sources: int | None = Field(default=None, ge=0)
    recovery_cache_captured: int | None = Field(default=None, ge=0)
    recovery_error_category: str | None = Field(default=None, max_length=64)
    ios_library_state: str | None = Field(default=None, max_length=32)
    ios_hidden_captured: int | None = Field(default=None, ge=0)
    ios_recently_deleted_captured: int | None = Field(default=None, ge=0)
    ios_cache_captured: int | None = Field(default=None, ge=0)
    ios_deleted_metadata_captured: int | None = Field(default=None, ge=0)
    ios_library_warning_count: int | None = Field(default=None, ge=0)
    analysis_scope: str | None = None
    device_sources: list[str] | None = None
    social_targets: list[str] | None = None
    report_sha256: str | None = None
    authorized_confirmed_findings: int | None = Field(default=None, ge=0)


class TimingBreakdown(ResponseModel):
    t_detect_ms: float = Field(default=0, ge=0)
    t_acquire_ms: float = Field(default=0, ge=0)
    t_inventory_ms: float = Field(default=0, ge=0)
    t_preprocess_ms: float = Field(default=0, ge=0)
    t_selection_ms: float = Field(default=0, ge=0)
    t_transfer_ms: float = Field(default=0, ge=0)
    t_index_ms: float = Field(default=0, ge=0)
    t_analyze_ms: float = Field(default=0, ge=0)
    t_total_ms: float = Field(default=0, ge=0)


class SessionSummary(ResponseModel):
    id: str
    device_id: str
    device_type: DeviceType
    label: str
    mode: AcquisitionMode
    scenario: Scenario
    status: SessionStatus
    progress: SessionProgress
    timing: TimingBreakdown
    participant: ParticipantIdentity | None = None
    recommendation: str | None = None
    review_candidates: bool = False
    created_at: str
    updated_at: str
    error: str | None = None


class PaginatedSessions(ResponseModel):
    items: list[SessionSummary] = Field(default_factory=list)
    page: int = Field(ge=1)
    page_size: int = Field(ge=1)
    total: int = Field(ge=0)
    pages: int = Field(ge=1)

File: frontend/src/shared/api/types/session.ts
import type {
  AcquisitionMode,
  AnalysisScope,
  ReviewStatus,
  Scenario,
  SessionStatus,
} from "./common";
import type { SocialPreview } from "./social";

export interface SessionProgress {
  phase: SessionStatus;
  percent: number;
  message: string;
  files_listed: number;
  files_pulled: number;
  files_indexed: number;
  files_analyzed: number;
  findings_count: number;
  throughput_files_per_sec: number;
  live_analysis_ms?: number;
  acquisition_method?: string | null;
  hits_l1?: number;
  hits_l2?: number;
  hits_l3?: number;
  hits_l4?: number;
  hits_ocr?: number;
  hits_asr?: number;
  whatsapp_state?: string | null;
  whatsapp_ui_attempt?: number;
  whatsapp_ui_attempts?: number;
  whatsapp_messages?: number;
  whatsapp_conversations?: number;
  whatsapp_parse_skipped?: number;
  notes_state?: string | null;
  notes_flow?: string | null;
  notes_app?: string | null;
  notes_captured?: number;
  notes_skipped?: number;
  notes_warning_count?: number;
  crawl_state?: string | null;
  crawl_source?: string | null;
  crawl_target?: string | null;
  crawl_scope?: string | null;
  crawl_stage?: string | null;
  crawl_attempt?: number | null;
  crawl_attempt_state?: string | null;
  crawl_failure_class?: string | null;
  crawl_reason?: string | null;
  crawl_scroll_count?: number | null;
  crawl_screenshot_count?: number | null;
  authorized_by?: string | null;
  authorized_at?: string | null;
  authorize_note?: string | null;
  analysis_scope?: AnalysisScope | null;
  device_sources?: string[];
  social_targets?: string[];
  report_sha256?: string | null;
  authorized_confirmed_findings?: number | null;
  crawl_discovered?: number;
  crawl_duplicates?: number;
  preprocessing_state?: string;
  preprocessing_total?: number;
  preprocessing_completed?: number;
  preprocessing_skipped?: number;
  preprocessing_truncated?: number;
  preprocessing_failed?: number;
  preprocessing_preprocessor_totals?: Record<
    string,
    {
      attempted?: number;
      processed?: number;
      skipped?: number;
      truncated?: number;
      failed?: number;
      cancelled?: number;
    }
  >;
  selection_evaluated?: number;
  selection_candidates?: number;
  selection_selected?: number;
  transfer_completed?: number;
  transfer_records?: number;
  transfer_artifacts?: number;
  android_inventory_ms?: number;
  android_preprocessing_ms?: number;
  android_selection_ms?: number;
  android_transfer_ms?: number;
  recovery_state?: "scanning" | "complete" | "partial" | "unavailable";
  recovery_mode?: AcquisitionMode;
  recovery_candidates?: number;
  recovery_captured?: number;
  recovery_bytes?: number;
  recovery_warning_count?: number;
  recovery_duration_ms?: number;
  recovery_cache_sources?: number;
  recovery_cache_captured?: number;
  recovery_error_category?: string;
  ios_library_state?: string;
  ios_hidden_captured?: number;
  ios_recently_deleted_captured?: number;
  ios_cache_captured?: number;
  ios_deleted_metadata_captured?: number;
  ios_library_warning_count?: number;
}

export interface TimingBreakdown {
  t_detect_ms: number;
  t_acquire_ms: number;
  t_inventory_ms?: number;
  t_preprocess_ms?: number;
  t_selection_ms?: number;
  t_transfer_ms?: number;
  t_index_ms: number;
  t_analyze_ms: number;
  t_total_ms: number;
}

export interface ParticipantIdentity {
  full_name: string;
  registration_no: string;
  nik?: string | null;
  organization?: string | null;
}

export interface SessionSummary {
  id: string;
  device_id: string;
  device_type: import("./common").DeviceType;
  label: string;
  mode: AcquisitionMode;
  scenario: Scenario;
  status: SessionStatus;
  progress: SessionProgress;
  timing: TimingBreakdown;
  participant?: ParticipantIdentity | null;
  recommendation: string | null;
  created_at: string;
  updated_at: string;
  error: string | null;
  review_candidates?: boolean;
}

export interface Finding {
  id: string;
  session_id: string;
  file_id: string;
  source: string;
  path: string;
  category: string;
  label: string;
  confidence: number;
  layer_origin: string;
  evidence: string;
  review_status: ReviewStatus;
  created_at: string;
  media_year?: number | null;
  media_captured_at?: string | null;
  preview_path?: string | null;
  preview_text?: string | null;
  source_app?: string | null;
  social_scope?: string | null;
  presentation?: "file" | "visual" | "text" | "chat";
  social_preview?: SocialPreview | null;
  reviewed_by?: string | null;
  reviewed_at?: string | null;
}
